### PR TITLE
Use `GroupNode` and `LeafNode` wrapper objects for both groups and leaf nodes

### PR DESCRIPTION
**Pull request by dimaryaz**
Makes things more consistent, and allows accessing leaf nodes without
having to load the dataframe into memory.

### CODE REVIEW

File: quilt/data.py [hunk 1]
```diff
@@ -79,6 +72,16 @@ def __dir__(self):
         # https://mail.python.org/pipermail/python-ideas/2011-May/010321.html
         return sorted(set((dir(type(self)) + list(self.__dict__) + self._keys())))
 
+    def _is_group(self):
+        return isinstance(self._node, GroupNode)
+
+    def _is_df(self):
+        return not isinstance(self._node, GroupNode)
+
+    def _df(self):
```
Comment (dimaryaz) [inline quilt/data.py]: I don't know if `_df()` is the way to go, though: we already have a `_dfs()` method which returns children paths, so it's confusing. Also, do we consider file nodes to be dataframes, too?

I want it to be concise, so not know what to call it. We could even do just `()`, though that's hacky.

Comment (akarve) [inline quilt/data.py]: `_df()` is OK for now. suggest `_dfkeys()` and `_groupkeys()` for the other methods; will need to `README.md` with new APIs

**Comment by akarve** [comment]
looks good but missing one thing: changes to README.md to reflect new APIs

File: quilt/data.py [hunk 2]
```diff
@@ -56,38 +54,43 @@ def __getattr__(self, name):
...
 
     def __dir__(self):
         # https://mail.python.org/pipermail/python-ideas/2011-May/010321.html
         return sorted(set((dir(type(self)) + list(self.__dict__) + self._keys())))
 
-    def _dfs(self):
+    def _is_group(self):
+        return isinstance(self._node, GroupNode)
+
+    def _is_df(self):
```
Comment (kevinemoore) [inline quilt/data.py]: We shouldn't report a node as a df unless it's actually a meaningful Pandas DF. Files, e.g., should not be dfs. Do we need a new concept, e.g., LeafNode that specifies merely that a node has no children?

Comment (dimaryaz) [inline quilt/data.py]: Yeah, that's what I was wondering. I can rename `df` to `leaf`

File: quilt/data.py [hunk 3]
```diff
@@ -31,12 +30,11 @@ class DataNode(object):
     in HDFStore's `root`.
```
Comment (dimaryaz) [inline quilt/data.py]: Oh, yes

**Comment by akarve** [comment]
Maybe DataNode === Leaf. because we only have Data at the leaves. Everything else is really a PackageNode or a GroupNode in my mind.

File: quilt/data.py (hunk 2 above)
Comment (dimaryaz) [inline quilt/data.py]: So technically, `_df()` doesn't make sense, either. `_get()`?

Comment (kevinemoore) [inline quilt/data.py]: No, _df() makes sense, but it should only apply to nodes that are DataFrames. It should either not be available or throw an exception for groups or files I would think.

**Comment by akarve** [comment]
blame me for `()`. i asked for it. it's just an alias for `data()`. as a quilt power user i want to type less.

**Pull request merged**